index.js

```diff
@@ -48,6 +48,7 @@ HtmlWebpackPlugin.prototype.apply = function (compiler) {
         // If the compilation change didnt change the cache is valid
         isCompilationCached = compilationResult.hash && self.childCompilerHash === compilationResult.hash;
         self.childCompilerHash = compilationResult.hash;
+        self.options.filename = self.options.filename.replace(/\[hash\]/g, compilationResult.hash);
```

Comment:
We could store the filename in a different property such as `this.modifiedFilename` and leave the option as is for clarity.